### ISSUE TITLE
Generated HTML <img> code is missing alt text (WCAG guideline)

### ISSUE THREAD

**Issue by weeheavy** [pelican-plugins/graphviz#30]
According to https://webaim.org/standards/wcag/checklist#sc1.1.1 (which is the abstract of https://www.w3.org/TR/WCAG22/#non-text-content), in this case `<img>` elements should have alt texts.

It's currently not possible to specify alt text within the Markdown's Graphviz code block which would then be passed resp. used as alt text in the `<img>` HTML element.

A simple code of
```graphviz
..graphviz dot
strict digraph foo {
"a" -> "b";
}
```

Results in this HTML output (no `alt=` attribute there):
```html
<div class="graphviz">
    <img src="data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+CjwhRE9DVFlQRSBzdmcgUFVCTElDICItLy9XM0MvL0RURCBTVkcgMS4xLy9FTiIKICJodHRwOi8vd3d3LnczLm9yZy9HcmFwaGljcy9TVkcvMS4xL0RURC9zdmcxMS5kdGQiPgo8IS0tIEdlbmVyYXRlZCBieSBncmFwaHZpeiB2ZXJzaW9uIDEyLjIuMSAoMjAyNDEyMDYuMjM1MykKIC0tPgo8IS0tIFRpdGxlOiBmb28gUGFnZXM6IDEgLS0+Cjxzdmcgd2lkdGg9IjYycHQiIGhlaWdodD0iMTE2cHQiCiB2aWV3Qm94PSIwLjAwIDAuMDAgNjIuMDAgMTE2LjAwIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIj4KPGcgaWQ9ImdyYXBoMCIgY2xhc3M9ImdyYXBoIiB0cmFuc2Zvcm09InNjYWxlKDEgMSkgcm90YXRlKDApIHRyYW5zbGF0ZSg0IDExMikiPgo8dGl0bGU+Zm9vPC90aXRsZT4KPHBvbHlnb24gZmlsbD0id2hpdGUiIHN0cm9rZT0ibm9uZSIgcG9pbnRzPSItNCw0IC00LC0xMTIgNTgsLTExMiA1OCw0IC00LDQiLz4KPCEtLSBhIC0tPgo8ZyBpZD0ibm9kZTEiIGNsYXNzPSJub2RlIj4KPHRpdGxlPmE8L3RpdGxlPgo8ZWxsaXBzZSBmaWxsPSJub25lIiBzdHJva2U9ImJsYWNrIiBjeD0iMjciIGN5PSItOTAiIHJ4PSIyNyIgcnk9IjE4Ii8+Cjx0ZXh0IHRleHQtYW5jaG9yPSJtaWRkbGUiIHg9IjI3IiB5PSItODQuOTUiIGZvbnQtZmFtaWx5PSJUaW1lcyxzZXJpZiIgZm9udC1zaXplPSIxNC4wMCI+YTwvdGV4dD4KPC9nPgo8IS0tIGIgLS0+CjxnIGlkPSJub2RlMiIgY2xhc3M9Im5vZGUiPgo8dGl0bGU+YjwvdGl0bGU+CjxlbGxpcHNlIGZpbGw9Im5vbmUiIHN0cm9rZT0iYmxhY2siIGN4PSIyNyIgY3k9Ii0xOCIgcng9IjI3IiByeT0iMTgiLz4KPHRleHQgdGV4dC1hbmNob3I9Im1pZGRsZSIgeD0iMjciIHk9Ii0xMi45NSIgZm9udC1mYW1pbHk9IlRpbWVzLHNlcmlmIiBmb250LXNpemU9IjE0LjAwIj5iPC90ZXh0Pgo8L2c+CjwhLS0gYSYjNDU7Jmd0O2IgLS0+CjxnIGlkPSJlZGdlMSIgY2xhc3M9ImVkZ2UiPgo8dGl0bGU+YSYjNDU7Jmd0O2I8L3RpdGxlPgo8cGF0aCBmaWxsPSJub25lIiBzdHJva2U9ImJsYWNrIiBkPSJNMjcsLTcxLjdDMjcsLTY0LjQxIDI3LC01NS43MyAyNywtNDcuNTQiLz4KPHBvbHlnb24gZmlsbD0iYmxhY2siIHN0cm9rZT0iYmxhY2siIHBvaW50cz0iMzAuNSwtNDcuNjIgMjcsLTM3LjYyIDIzLjUsLTQ3LjYyIDMwLjUsLTQ3LjYyIi8+CjwvZz4KPC9nPgo8L3N2Zz4K">
</div>
```

Decoding the Base64, a `<title>foo</title>` is there which probably could be used as the source for this, but that's obviously the SVG title itself.

Software versions used:
- graphviz version 12.2.1 (20241206.2353)
- Pelican graphviz 1.2.5
- Pelican 4.10.2

**Comment by rlaboiss** [comment]
Thank you for your suggestion.

> A simple code of
> 
> ```
> ..graphviz dot
> strict digraph foo {
> "a" -> "b";
> }
> ```
> 
> Results in this HTML output (no `alt=` attribute there):
> 
> ```
> <div class="graphviz">
>     <img src="data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+CjwhRE9DVFlQRSBzdmcgUFVCTElDICItLy9XM0MvL0RURCBTVkcgMS4xLy9FTiIKICJodHRwOi8vd3d3LnczLm9yZy9HcmFwaGljcy9TVkcvMS4xL0RURC9zdmcxMS5kdGQiPgo8IS0tIEdlbmVyYXRlZCBieSBncmFwaHZpeiB2ZXJzaW9uIDEyLjIuMSAoMjAyNDEyMDYuMjM1MykKIC0tPgo8IS0tIFRpdGxlOiBmb28gUGFnZXM6IDEgLS0+Cjxzdmcgd2lkdGg9IjYycHQiIGhlaWdodD0iMTE2cHQiCiB2aWV3Qm94PSIwLjAwIDAuMDAgNjIuMDAgMTE2LjAwIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIj4KPGcgaWQ9ImdyYXBoMCIgY2xhc3M9ImdyYXBoIiB0cmFuc2Zvcm09InNjYWxlKDEgMSkgcm90YXRlKDApIHRyYW5zbGF0ZSg0IDExMikiPgo8dGl0bGU+Zm9vPC90aXRsZT4KPHBvbHlnb24gZmlsbD0id2hpdGUiIHN0cm9rZT0ibm9uZSIgcG9pbnRzPSItNCw0IC00LC0xMTIgNTgsLTExMiA1OCw0IC00LDQiLz4KPCEtLSBhIC0tPgo8ZyBpZD0ibm9kZTEiIGNsYXNzPSJub2RlIj4KPHRpdGxlPmE8L3RpdGxlPgo8ZWxsaXBzZSBmaWxsPSJub25lIiBzdHJva2U9ImJsYWNrIiBjeD0iMjciIGN5PSItOTAiIHJ4PSIyNyIgcnk9IjE4Ii8+Cjx0ZXh0IHRleHQtYW5jaG9yPSJtaWRkbGUiIHg9IjI3IiB5PSItODQuOTUiIGZvbnQtZmFtaWx5PSJUaW1lcyxzZXJpZiIgZm9udC1zaXplPSIxNC4wMCI+YTwvdGV4dD4KPC9nPgo8IS0tIGIgLS0+CjxnIGlkPSJub2RlMiIgY2xhc3M9Im5vZGUiPgo8dGl0bGU+YjwvdGl0bGU+CjxlbGxpcHNlIGZpbGw9Im5vbmUiIHN0cm9rZT0iYmxhY2siIGN4PSIyNyIgY3k9Ii0xOCIgcng9IjI3IiByeT0iMTgiLz4KPHRleHQgdGV4dC1hbmNob3I9Im1pZGRsZSIgeD0iMjciIHk9Ii0xMi45NSIgZm9udC1mYW1pbHk9IlRpbWVzLHNlcmlmIiBmb250LXNpemU9IjE0LjAwIj5iPC90ZXh0Pgo8L2c+CjwhLS0gYSYjNDU7Jmd0O2IgLS0+CjxnIGlkPSJlZGdlMSIgY2xhc3M9ImVkZ2UiPgo8dGl0bGU+YSYjNDU7Jmd0O2I8L3RpdGxlPgo8cGF0aCBmaWxsPSJub25lIiBzdHJva2U9ImJsYWNrIiBkPSJNMjcsLTcxLjdDMjcsLTY0LjQxIDI3LC01NS43MyAyNywtNDcuNTQiLz4KPHBvbHlnb24gZmlsbD0iYmxhY2siIHN0cm9rZT0iYmxhY2siIHBvaW50cz0iMzAuNSwtNDcuNjIgMjcsLTM3LjYyIDIzLjUsLTQ3LjYyIDMwLjUsLTQ3LjYyIi8+CjwvZz4KPC9nPgo8L3N2Zz4K">
> </div>
> ```
> 
> Decoding the Base64, a `<title>foo</title>` is there which probably could be used as the source for this, but that's obviously the SVG title itself.

This is the SVG content generated by your example:

```svg
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 12.2.1 (20241206.2353)
 -->
<!-- Title: foo Pages: 1 -->
<svg width="62pt" height="116pt"
 viewBox="0.00 0.00 62.00 116.00" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 112)">
<title>foo</title>
<polygon fill="white" stroke="none" points="-4,4 -4,-112 58,-112 58,4 -4,4"/>
<!-- a -->
<g id="node1" class="node">
<title>a</title>
<ellipse fill="none" stroke="black" cx="27" cy="-90" rx="27" ry="18"/>
<text text-anchor="middle" x="27" y="-84.95" font-family="Times,serif" font-size="14.00">a</text>
</g>
<!-- b -->
<g id="node2" class="node">
<title>b</title>
<ellipse fill="none" stroke="black" cx="27" cy="-18" rx="27" ry="18"/>
<text text-anchor="middle" x="27" y="-12.95" font-family="Times,serif" font-size="14.00">b</text>
</g>
<!-- a&#45;&gt;b -->
<g id="edge1" class="edge">
<title>a&#45;&gt;b</title>
<path fill="none" stroke="black" d="M27,-71.7C27,-64.41 27,-55.73 27,-47.54"/>
<polygon fill="black" stroke="black" points="30.5,-47.62 27,-37.62 23.5,-47.62 30.5,-47.62"/>
</g>
</g>
</svg>
```

Notice that there are several `<title>*</title>` instances, one for each Graphviz object. However, there is this line:
```
<!-- Title: foo Pages: 1 -->
```
that clearly indicates the ID of the top level Graphviz object. This could indeed be used for the `alt` property of the `img` element.

However, note that it is totally valid to omit the ID in the dot language, like this:
```
strict digraph {
"a" -> "b";
}
```

In this case, the Tilte (and hence the `alt` property) will be empty.

We might also allow the user to specify the `alt` text in Markdown, using the configuration mechanism of the `..graphviz` block, which would override the graph title, something like this:

```
..graphviz [alt-text="Wonderful graph"] dot
strict digraph foo {
"a" -> "b";
}
```

What do you think?

**Comment by weeheavy** [comment]
I agree, the `<title>*</title>` elements are not suitable for a broad alt text. I very much like your idea to specify the alt text in the main `..graphviz` area. I think as long as this would be optional for people having no use for it, that could be the way to go.

**Comment by rlaboiss** [comment]
Yes, the idea is to have `alt-text` optional. If the user uses it, then the `alt` property will be set accordingly. If the `alt-text` option is absent, then the ID of the top-level Graphviz will be used (`foo`, in your example). Otherwise, if there is no ID, then a default `alt` text will be generated (possibly controlled by a plugin configuration variable, like `GRAPHVIZ_ALT_TEXT`).

I will implement this as soon as I will have some free time. Thanks again for your suggestion and for the discussion.

**Comment by rlaboiss** [comment]
I am hereby closing this issue, since it should be fixed in release 1.3.0 of pelican-graphviz.

@weeheavy: Could you please test this release? If you find a problem, please open a new issue.

**Comment by rlaboiss** [comment]
P.S.: @weeheavy: I forgot to tell that you are [acknowledged](https://github.com/pelican-plugins/graphviz?tab=readme-ov-file#acknowledgments) in the `README.md` file.

**Comment by weeheavy** [comment]
It works flawlessly. Thank you very much for the fast implementation and the kind interaction.
